lib/agent0/agent0/hyperdrive/interactive/i_hyperdrive.py

```diff
@@ -0,0 +1,532 @@
+"""Defines the interactive hyperdrive class that encapsulates a hyperdrive pool."""
+
+from __future__ import annotations
+
+import asyncio
+import logging
+import os
+from dataclasses import asdict, dataclass
+from typing import Any, Literal, Type, overload
+
+import nest_asyncio
+import numpy as np
+from eth_account.account import Account
+from eth_account.signers.local import LocalAccount
+from ethpy import EthConfig
+from ethpy.base import set_anvil_account_balance, smart_contract_transact
+from ethpy.hyperdrive import (
+    HyperdriveAddresses,
+    HyperdriveReadWriteInterface,
+    ReceiptBreakdown,
+    fetch_hyperdrive_address_from_uri,
+)
+from fixedpointmath import FixedPoint
+from numpy.random._generator import Generator
+from web3 import Web3
+
+from agent0.hyperdrive import HyperdriveActionType, HyperdriveAgent, TradeResult, TradeStatus
+from agent0.hyperdrive.agent import build_wallet_positions_from_chain
+from agent0.hyperdrive.crash_report import log_hyperdrive_crash_report
+from agent0.hyperdrive.exec import async_execute_agent_trades, set_max_approval
+from agent0.hyperdrive.policies import HyperdriveBasePolicy
+from agent0.test_utils import assert_never
+
+from .event_types import (
+    AddLiquidity,
+    CloseLong,
+    CloseShort,
+    OpenLong,
+    OpenShort,
+    RedeemWithdrawalShares,
+    RemoveLiquidity,
+)
+from .i_chain import IChain
+from .i_hyperdrive_agent import IHyperdriveAgent
+from .i_hyperdrive_policy import IHyperdrivePolicy
+
+# In order to support both scripts and jupyter notebooks with underlying async functions,
```

Comment:
neat! I am ok with leaving this in here. For our use case, being able to hot swap between jupyter and scripts is worth it. Can you add something like

```py
# TODO: nest_asyncio may cause compatibility issues with other libraries.
# Also, Jupyter and ASYNC compatibility might be improved, removing the need for this.
# See https://github.com/python/cpython/issues/66435.
```

I read [this python issue](https://github.com/python/cpython/issues/66435) about it. Honestly, the discussion was huge and had lots of rants with some more spicy personal stuff sprinkled in. A good read if you're looking to dive into some python dev drama 😆 . You definitely don't have to slog through it, but I want to bring your attention to some notable dissension (from Guido) to using supporting nested async loops or async->sync->async pipelines:

> In my own experience, the very programming technique that you are proposing has caused some very hard to debug problems that appeared as very infrequent and hard to predict stack overflows.

 >  asyncio has a guarantee that no other tasks run unless await is used. This creates an implicit "critical section" between any two await calls.
 
 > [...] no blocking I/O should be performed (all I/O must be non-blocking and defer to the event loop using await). The reason for this requirement is that such blocking I/O calls would prevent the I/O parallelism promised by the asyncio architecture -- the blocking call [...] prevents all other asynchronous I/O from happening.
 
I am immediately thinking of web3, which I believe is implementing the "two libraries" approach, with normal web3 and a largely duplicated async web3 that includes async/await calls throughout. We want to take care to monitor for edge cases where our use of async (or "hammers" on top of, like `nest_asyncio`), causes incompatibility issues. For example in the situation of running agent0, which wraps web3, in a jupyter notebook. Hopefully our testing sweet will catch this, and also there seems to be some momentum to fix the Jupyter & asyncio compatibility soon anyway.